replica/table.cc

```diff
@@ -2837,4 +2838,210 @@ bool table::requires_cleanup(const sstables::sstable_set& set) const {
     }));
 }
 
+unsigned compaction_group::sstable_set::x_log2_compaction_groups(size_t number_of_cgs) {
+    if (number_of_cgs & (number_of_cgs - 1)) {
+        on_internal_error(tlogger, format("Number of compaction groups ({}) is not aligned to power-of-two.", number_of_cgs));
+    }
+    return std::log2(number_of_cgs);
+}
+
+compaction_group::sstable_set::sstable_sets_t
+compaction_group::sstable_set::initialize_sstable_sets(const std::vector<std::unique_ptr<compaction_group>>& cgs) const {
+    if (cgs.empty()) {
+        on_internal_error(tlogger, "Compaction group SSTable set cannot be initialized with a empty list.");
+    }
+    return boost::copy_range<sstable_sets_t>(cgs | boost::adaptors::transformed([] (const std::unique_ptr<compaction_group>& cg) {
+        return std::make_pair(cg->token_range(), cg->make_compound_sstable_set());
+    }));
+}
+
+unsigned compaction_group::sstable_set::group_of(const sstable_sets_t& sets, const dht::token& t) {
+    auto idx = dht::compaction_group_of(x_log2_compaction_groups(sets.size()), t);
+    if (idx >= sets.size()) {
+        throw std::out_of_range(format("Token {} was mapped to an nonexistent compaction group {}. Valid range [0..{}].",
+                                       t, idx, sets.size() - 1));
+    }
+    return idx;
+}
+
+unsigned compaction_group::sstable_set::group_of(const dht::token& t) const {
+    return group_of(_cg_sstable_sets, t);
+}
+
+std::pair<unsigned, unsigned> compaction_group::sstable_set::groups_of(const dht::token_range& tr) const {
+    auto start = group_of(tr.start() ? tr.start()->value() : dht::minimum_token());
+    auto end = group_of(tr.end() ? tr.end()->value() : dht::maximum_token());
```

Comment:
Need to consider inclusive and exclusive bounds. This can be important when selecting exactly one tablet. Please add tests for this.